### PR TITLE
Make index_add_ error if input source shape is wrong

### LINKED ISSUE #92576: `index_add_` should error if shape of source is wrong
### 🐛 Describe the bug

`index_add_` will fail in the backward-mode gradient computation. It should succeed without any exception

```py
import torch
b = torch.tensor([2., 4.])

def func(b):
    a = torch.tensor([[1., 2.],
                        [4., 5.],
                        [7., 8.]])
    a.index_add_(0, torch.tensor([0, 2]), b)
    return a

ans = func(b.clone().requires_grad_())
ans.sum().backward()
```

```
RuntimeError: expand(torch.FloatTensor{[2, 2]}, size=[2]): the number of sizes provided (1) must be greater or equal to the number of dimensions in the tensor (2)
```

### Versions

```
PyTorch version: 2.0.0.dev20230105
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.15 (main, Nov 24 2022, 14:31:59)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090

Nvidia driver version: 515.86.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.0.dev20230105
[pip3] torchaudio==2.0.0.dev20230105
[pip3] torchvision==0.15.0.dev20230105
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.23.5           py39h14f4228_0
[conda] numpy-base                1.23.5           py39h31eccc5_0
[conda] pytorch                   2.0.0.dev20230105 py3.9_cuda11.7_cudnn8.5.0_0    pytorch-nightly
[conda] pytorch-cuda              11.7                 h67b0de4_2    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchaudio                2.0.0.dev20230105      py39_cu117    pytorch-nightly
[conda] torchtriton               2.0.0+0d7e753227            py39    pytorch-nightly
[conda] torchvision               0.15.0.dev20230105      py39_cu117    pytorch-nightly
```

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by tfsingh**
Fixes #92576 , checking the following as described in the documentation:

"source.shape[dim] == len(index) and source.shape[i] == self.shape[i] for i != dim"

Would be happy to iterate on this if there are any issues, and would be happy to implement the checking for the CUDA and MPS implementations of index_add_.

### CODE REVIEW

**Comment by tfsingh** [comment]
The test fallback_masked_fill might need to be changed -- the implementation looks roughly something like this (excluding the batching using vmap).

```
import torch
x = torch.randn(7, 11, 13)
dim = 0
index = torch.tensor([0, 4, 2])
values = torch.randn(3, 13)
ans = x.index_add(dim, index, values.clone().requires_grad_())
```

However, when we compute the backward mode gradient computation, we see the same error as described in the original issue (all done in a fresh colab notebook).

```
ans.sum().backward()
RuntimeError: expand(torch.FloatTensor{[3, 11, 13]}, size=[3, 13]): the number of sizes provided (2) must be greater or equal to the number of dimensions in the tensor (3)
```

Since we'd like to do this check in the forward, I think we ought to change this test to expect the runtime error this pr throws. I suspect whatever's failing in the onnx tests is similar, so I'll look at those too.

Would greatly appreciate your input @lezcano (apologies as well for the previous pr which was botched, but I'm dedicated to getting this issue fixed).

**Comment by lezcano** [comment]
Those errors are failing in CUDA. You modified the cpu version of the function. Now, you just need to modify accordingly the CUDA version of the function. You can either repeat the code or have a function that's shared between the two. Many functions have the checks shared, but yeah, this one repeats them, which is not great.

**Comment by tfsingh** [comment]
Okay, just added -- hoping this works, but am of course happy to iterate on it if not.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp]: Perhaps you meant `source.size(dim) == numel`?

Comment (lezcano) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: Also, try running the tests locally before pushing when possible to save on CI runs.

Comment (tfsingh) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: Yes 100%, I'm not sure how my initial changes were erased -- I still thought it might make sense to remove it as the same check is done on line ~335 in index_func_meta_impl.
```
TORCH_CHECK(numel == (source.dim() == 0 ? 1 : source.size(dim)),
              func, "_(): Number of indices (", numel, ") should be equal to source.size(dim): (",
              source.size(dim), "), for dim: ", dim)
```
Do you think I should reinclude the removed check?

Comment (lezcano) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: Then, yeah, don't add this check. But also, add the checks in `index_func_meta_impl` that way we can avoid the repetition.

Comment (tfsingh) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: > Then, yeah, don't add this check. But also, add the checks in `index_func_meta_impl` that way we can avoid the repetition.

That makes sense, after checking the documentation it seems like [reduce](https://pytorch.org/docs/stable/generated/torch.Tensor.index_reduce_.html)/[select](https://pytorch.org/docs/stable/generated/torch.index_select.html)/etc require the same.

Comment (lezcano) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: you should have a look at which tests are failing and then run `pytest test/whichever_file_is_the_test_in.py -vk name_of_test`

**Comment by tfsingh** [comment]
> The test fallback_masked_fill might need to be changed -- the implementation looks roughly something like this (excluding the batching using vmap).
> 
> ```
> import torch
> x = torch.randn(7, 11, 13)
> dim = 0
> index = torch.tensor([0, 4, 2])
> values = torch.randn(3, 13)
> ans = x.index_add(dim, index, values.requires_grad_()) # I appended the requires gradient as the error wouldn't be thrown without it, even though it should as described in the documentation
> ```
> 
> However, when we compute the backward mode gradient computation, we see the same error as described in the original issue (all done in a fresh colab notebook).
> 
> ```
> ans.sum().backward()
> RuntimeError: expand(torch.FloatTensor{[3, 11, 13]}, size=[3, 13]): the number of sizes provided (2) must be greater or equal to the number of dimensions in the tensor (3)
> ```
> 
> Since we'd like to do this check in the forward, I think we ought to change this test to expect the runtime error this pr throws. I suspect whatever's failing in the onnx tests is similar, so I'll look at those too.
> 
> Would greatly appreciate your input @lezcano (apologies as well for the previous pr which was botched, but I'm dedicated to getting this issue fixed).

Re: failing tests. I think the remaining ones still fall under this analysis, in that they violate the invariant we're checking for. In the case of fallback_max_fill, we are given a self of shape (7,11,13) and source (3, 13). Omitting self[dim] and source[dim] gives shapes (11,13) and (13), which aren't broadcastable. This issue exists for the onnx tests as well, where we try to broadcast shape of (5,4,3) with shape (1,3,3) along dim = 1 in index_add_dim_size_differ and shape (5,4,3) with shape (1,3,3) along dim = 1 in index_add_dynamic_axes.

I'm not sure if we should rewrite them to throw the expected error, or if there's a better course of action here.

**Comment by lezcano** [comment]
>gives shapes (11,13) and (13), which aren't broadcastable. 
These shapes are broadcastable. This means that the check needs to be a bit more generic to account for broadcasting.

**Review by lezcano — COMMENTED** [review]
The changes should just be done in the meta impl. The meta impl is common to both the CPU and CUDA parts

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: This is not the check that we want. This does not check for broadcasting of any kind.

File: test/test_torch.py
Comment (lezcano) [inline test/test_torch.py:5955]: `test_torch.py` is legacy. Please add an error input the relevant OpInfo file in `torch/testing/_internal/common_methods_invocations.py`

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (tfsingh) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: I may be mistaken, but I believe infer_size_impl in ExpandUtils.cpp checks for broadcasting before reaching TensorAdvancedIndexing. Therefore, I believe this should check the only potential failure case with dimensionality, for if we reach this point in the program flow, we can assume the shapes are broadcastable.

If I'm incorrect, I'll need help on how to vectorize the check for broadcastability as you requested earlier (in which case we should update the code for infer_size_impl as it checks using a for loop).

Comment (lezcano) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: This function is called form 
```c++
TORCH_PRECOMPUTE_META_FUNC(index_add)
(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& source, const Scalar& alpha) {
  dim = maybe_wrap_dim(dim, self.dim());
  index_func_meta_impl(*this, self, dim, index, source, "index_add");
  return TORCH_PRECOMPUTE_STRUCT(index_add)().set_dim(dim);
}
```
(among other places), so there is no check here that source has the correct size.

You don't need to implement the broadcasting check yourself. There are utility functions to do this. I don't remember the name of them though, but shouldn't be difficult to find.

File: test/test_torch.py
Comment (tfsingh) [reply test/test_torch.py]: Just moved it over, please let me know if I should iterate on my changes.

**Comment by tfsingh** [comment]
I'm unsure as to why the error is not being raised on certain devices, and how "source tensor shape must match self tensor shape (excluding the specified dimension)" does not match "source tensor shape must match self tensor shape (excluding the specified dimension)".

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:345]: Why creating a tensor just to call `.sizes()`? Pass the sizes directly.

Comment (tfsingh) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:345]: I tried this initially, but it wasn't working. ```test/test_legacy_vmap.py TestVmapAPI.test_fallback_masked_fill``` errors out when I try running it locally if passing the self_sizes and source_sizes directly (after erasure).

My assumption is that we need to call .vec() to erase the given dimension, and in order to convert back to the .sizes() representation, we must create tensors that we can call .sizes() on. This assumption might be faulty but I'm unsure how to transition from the vector representations back to IntArrayRefs is_expandable_to accepts otherwise.

**Comment by lezcano** [comment]
The changes LGTM so far. Now there are a number of other tests that are failing. These are failing because linked to a given function there are many other functions that are implemented in terms of it (e.g. those associated with `vmap(fn)`, the decomposition associated with `torch.compile`, and perhaps a few others). The way to go about this is:
- First, fix the `test_ops.py -vk test_errors_index_add` tests. That's related with the way you wrote the regex (perhaps it's too strict)
- Then, you can try to find the other function associated with the other tests. In order to find them:
  -  `git grep [name of function]` your way through the code base
  -  Read the tests and run through them with a debugger, to see which code is executed. 
- Once you find a relevant function (e.g. the batching rule that makes `vmap(index_add)` work) it should be fairly easy to fix it by making it throw the same error

Again, run the tests locally and make sure they pass before pushing if possible.

Good luck!

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (tfsingh) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: I tried converting the sizes into IntArrayRefs with a constructor, but this wasn't working well. Is there a better way to go from a vector representation to the one calling .sizes() on a tensor produces?

**Comment by tfsingh** [comment]
Thank you for your review. I've fixed the regex issue, and am having some trouble with the broadcasting check (see comment above). 

I've been looking into the batching rule fix, and I've narrowed it down to index_add_batch_rule in BatchRulesScatterOps. I've been erasing the batching dimension of self and other along with the given dimension and then conducting a check for broadcastability, but this doesn't seem to work. Erasing just the given dimension doesn't trigger the error. I've added an xfail to catch the bugs for this CI run after seeing the same existed for index_put, perhaps this approach is acceptable. 

Finally, I might need to go to this Friday's office hours as my run_test.py isn't showing the same errors as the CI -- I'm trying to be mindful of CI runs, but feel as if my feedback loop is faulty without them.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (lezcano) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:345]: if you just pass the vectors, they should be casted automatically to `IntArrayRef`s via the implicit constructor.

**Comment by lezcano** [comment]
You shouldn't run `run_test.py`. You should look at the failing test, have a look at its name and the file it's executed from in the given CI run and do as described in https://github.com/pytorch/pytorch/pull/100321#discussion_r1184121195.

**Comment by tfsingh** [comment]
These changes make the tests pass, and are in line with what I saw for the other index_* methods. I added an additional check as just checking for broadcastability wasn't correctly throwing errors.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: I don't think this logic is correct. It looks like it's crafted to make the tested example pass. You are literally asserting that either self is of dimension 1, or sources is of dimension greater than 1 or the sizes of source are NOT broadcastable to the sizes of self...

Comment (tfsingh) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: If the error in logic you're identifying is with the check for broadcastability, you shouldn't be applying DeMorgan's law to this portion as it's not in the negated expression. I had to add the negated expression as the error wasn't being thrown for the case self isn't empty and the source is (which I believe the first portion accurately checks).

Comment (lezcano) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: Buf a 0D tensor (which is not empty) is broadcastable to any size, so why would that have to throw an error?

Comment (lezcano) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: (and yes, I misread the parens, sorry for that)

Comment (tfsingh) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: > Buf a 0D tensor (which is not empty) is broadcastable to any size, so why would that have to throw an error?

I don't have a great answer, the cop-out is that we require this check to throw the error properly (I understand that this might be reasoning too heavily on abstractions). I think the answer to this has to do with the IntArrayRef representation from implicit construction differing from creating tensors and calling .sizes() on them, as the error was correctly thrown in the latter case. More broadly, perhaps the correct semantics for index_* operations differ slightly from broadcasting? (I'm significantly less sure of this though).

If you have something you'd like me to change the check to I'd be happy to, but I'm not sure what that might be given you'd like me to pass the sizes directly.

Comment (lezcano) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:344]: The semantics of `index_*` operations are the same as that of `x[i] = val`, that is, you should be able to pass a value `val` that is broadcastable to the slice `x[i]`.

**Comment by lezcano** [comment]
More generally, I think it's alright to not fix the ref error, but I believe we need to fix the batching issue rather than just skipping it. I suggest you run with `gdb` through the given test and find where's the batching rule that's failing. Then, it shouldn't be too difficult to figure out how to fix it just by looking at how similar functions do it and looking at [the relevant documentation](https://github.com/pytorch/pytorch/blob/master/functorch/writing_batching_rules.md). If after some trying you still struggle with this, we can ask one of the functorch folks to help with this.

**Comment by tfsingh** [comment]
Okay, I looked at the relevant documentation and method, index_add_batch_rule in BatchRulesScatterOps.cpp. The batch rule augments self ([[1,2], [4,5], [7,8]]) into a tensor of size 2x3x2, ([[[1,2], [4,5], [7,8]], [[1,2], [4,5], [7,8]]]) and applies the same to source ([2,4] -> [[2,4],[2,4]]. Ignoring dimension 1 (the batching rule increments the dimension passed in by one to account for the batched portion) of these tensors gives us shapes of size (2,2) and (2). These shapes are broadcastable, so I don't see why they _should_ raise a runtime error as the test expects it to.

I think this goes back to the original discussion on the check itself -- I think the act of batching is incompatible with the edge case we're trying to catch, but I'm becoming gradually less sure if the case we're checking for is valid or what exactly the valid case is.

> Buf a 0D tensor (which is not empty) is broadcastable to any size, so why would that have to throw an error?

You make a good point -- I'm not exactly sure why we do here? Consider the case -- we're given indices [0,2], self [[1.,2.], [4.,5.], [7.,8.]], source [2., 4.], and dim=0. Thus, we should add 2. (index 0 of source) to index 0 of self ([1,2]) and 4. (index 1 of source) to index 2 of self [7,8]. That, to my understanding, is a fair operation and shouldn't error out? (this diligence was definitely worth doing before opening the PR, apologies for that). Perhaps there's something unique that should make this fail when we want gradients, but that's something I'm less familiar with. 

Regardless, this PR is taking quite a bit of time for both of us, I think this might be one of the few cases where a quick call might actually be helpful. I can reach out on the PyTorch Slack if you're open to this idea. If not, I understand. I don't quit, so I'll be waiting to hear your reply and will continue to make corrections until this is merged.

**Comment by lezcano** [comment]
sure, ping me over slack

**Review by lezcano — APPROVED** [review]
Great! Thank you for this one.

**Pull request closed without merging**
